### PR TITLE
Distinct last positional argument hashes from keywords on Ruby 3

### PR DESCRIPTION
**Pull request by JonRowe**
This is a solution attempt for rspec/rspec-support#536 

On Ruby 3 we can (with effort) distinguish between a last positional hash and a keyword hash, because the mixed behaviour we can get on other Rubies shouldn't happen.

### CODE REVIEW

File: lib/rspec/support/method_signature_verifier.rb [hunk 1]
```diff
@@ -85,12 +85,58 @@ def has_kw_args_in?(args)
             (args.last.empty? || args.last.keys.any? { |x| x.is_a?(Symbol) })
         end
 
-        # Without considering what the last arg is, could it
```
Comment (pirj) [inline lib/rspec/support/method_signature_verifier.rb:88]: I have a faint feeling that the incorrect behaviour is in the `has_kw_args_in?` above, not related to the `could_contain_kw_args?` call.

With the test case that reproduces the problem from #522,
```ruby
            it 'treats the final positional argument as a hash' do
              expect(valid?(1, { foo: 'bar', baz: 'eggs'})).to eq(true)
            end
```

The problem is a call to `arity_kw` defined as:
```ruby
def arity_kw(x, y = {}, z:2); end
```
 - **may** contain kwargs (so arguably `could_contain_kw_args?` returning `true` is correct)
 - has a variable number of positional arguments

The last argument can be matched to either `y` or considered a kwargs depending on what this last argument is.

With `{ foo: 'bar', baz: 'eggs'}` specifically, we can really detect it's a kwarg by checking if:
1. it is a Hash
2. all of its keys are Symbols
3. **all** of its keys of the arg are in the method signature as kwargs

With `foo` and `bar` as the last arg keys and `z` as the only kwarg in the method signature, we can certainly tell that the last argument is **not** a kwarg.

File: lib/rspec/support/method_signature_verifier.rb [hunk 2]
```diff
@@ -85,12 +85,58 @@ def has_kw_args_in?(args)
...
+            last_arg_matches_any_keywords = !(args.last.keys & @allowed_kw_args).empty?
+            has_extra_keys = !(args.last.keys - @allowed_kw_args).empty?
+
+            # If we allow the keyword splat, and the last hash contains
+            # a specified keyword, then this hash contains keywords
+            return true if @allows_any_kw_args && last_arg_matches_any_keywords
+
+            # If we don't allow the keyword splat, and the last hash contains extra keys
+            # then it is a positional hash.
+            return false if !@allows_any_kw_args && has_extra_keys
```
Comment (pirj) [inline lib/rspec/support/method_signature_verifier.rb:125]: I guess there's a problem between 2.7 and 3.0+ with (`spec/rspec/support/method_signature_verifier_spec.rb:372`)
```ruby
> def arity_kw(x, y = {}, z:2); end

> arity_kw(nil, {a: 1}, {z: 2})
```
it works fine on 2.7 but fails on 3.0+:
```
ArgumentError: wrong number of arguments (given 3, expected 1..2)
```

To me, it seems that this expectation is not justified on Ruby 2.7:
```
expect(valid?(nil, :a => 1)).to eq(false)
```
as `y` is initialized with `{:a => 1}`.

I didn't look at other few failing examples yet.

File: lib/rspec/support/method_signature_verifier.rb (hunk 1 above)
Comment (pirj) [reply lib/rspec/support/method_signature_verifier.rb:88]: Sorry for the lengthy comment, I've tried to refresh my memory by writing things down. Glancing over, I see that your implementation is addressing this exact problem in `has_extra_keys`.

Comment (JonRowe) [reply lib/rspec/support/method_signature_verifier.rb:88]: Yep this PR is entirely about addressing ambiguity where traditional style hashes can be keywords when there is this duality.

**Comment by JonRowe** [comment]
Migrated to the monorepo as rspec/rspec#120

**Pull request closed without merging**